src/test/scala/com/databricks/spark/redshift/RedshiftSourceSuite.scala

```diff
@@ -231,20 +233,50 @@ class RedshiftSourceSuite
     }
   }
 
-  test("DefaultSource serializes data as Avro, then sends Redshift COPY command") {
+  test("DefaultSource using 'query' supports user schema, pruned and filtered scans") {
+
+    val params = Map(
+      "url" -> "jdbc:postgresql://foo/bar",
+      "tempdir" -> tempDir,
+      "query" -> "select * from test_table"
+    )
+
+    val jdbcWrapper = prepareUnloadTest(params)
+    val testSqlContext = new SQLContext(sc)
+
+    // Construct the source with a custom schema
+    val source = new DefaultSource(jdbcWrapper)
+    val relation = source.createRelation(testSqlContext, params, TestUtils.testSchema)
+
+    // Define a simple filter to only include a subset of rows
+    val filters: Array[Filter] =
+      Array(EqualTo("testBool", true),
+        EqualTo("testString", "Unicode是樂趣"),
+        GreaterThan("testDouble", 1234152.123124981),
+        LessThan("testDouble", Double.MaxValue),
+        GreaterThanOrEqual("testFloat", 1.0f),
+        LessThanOrEqual("testInt", 42))
+    val rdd = relation.asInstanceOf[PrunedFilteredScan].buildScan(Array("testByte", "testBool"), filters)
+
+    // We should now only have one matching row, with two columns
+    val filteredExpectedValues = Array(Row(1, true))
+    rdd.collect() zip filteredExpectedValues foreach {
+      case (loaded, expected) => loaded shouldBe expected
```

Comment:
Nit: I don't like should matchers. Also, it would be nice to use something similar to Spark SQL's `checkAnswer` so that you get useful debugging output for failures.